pkg/experiment/symbolizer/symbolizer.go

```diff
@@ -0,0 +1,448 @@
+
+		if !mappingsToSymbolize[loc.MappingId] {
+			continue
+		}
+
+		locsByMapping[loc.MappingId] = append(locsByMapping[loc.MappingId], locToSymbolize{
+			idx: i,
+			loc: loc,
+		})
+	}
+
+	return locsByMapping, nil
+}
+
+// symbolizeLocationsForMapping symbolizes a single mapping group
+func (s *Symbolizer) symbolizeLocationsForMapping(ctx context.Context, profile *googlev1.Profile, mappingID uint64, locs []locToSymbolize) error {
+	mapping := profile.Mapping[mappingID-1]
+
+	binaryName, err := s.extractBinaryName(profile, mapping)
+	if err != nil {
+		return err
+	}
+
+	buildID, err := s.extractBuildID(profile, mapping)
+	if err != nil {
+		return err
+	}
+
+	if buildID == "" {
+		return nil
+	}
+
+	req := s.createSymbolizationRequest(binaryName, buildID, mapping, locs)
+
+	if err := s.symbolize(ctx, &req); err != nil {
+		return err
+	}
+
+	s.updateProfileWithSymbols(profile, mapping, locs, req.locations)
+
+	return nil
+}
+
+// extractBinaryName extracts the binary name from the mapping
+func (s *Symbolizer) extractBinaryName(profile *googlev1.Profile, mapping *googlev1.Mapping) (string, error) {
+	if mapping.Filename < 0 || int(mapping.Filename) >= len(profile.StringTable) {
+		return "", fmt.Errorf("invalid mapping: filename index %d out of range (string table length: %d)",
+			mapping.Filename, len(profile.StringTable))
+	}
+
+	fullPath := profile.StringTable[mapping.Filename]
+	return filepath.Base(fullPath), nil
+}
+
+// extractBuildID extracts and sanitizes the build ID from the mapping
+func (s *Symbolizer) extractBuildID(profile *googlev1.Profile, mapping *googlev1.Mapping) (string, error) {
+	buildID := profile.StringTable[mapping.BuildId]
+	sanitizedBuildID, err := sanitizeBuildID(buildID)
+	if err != nil {
+		level.Error(s.logger).Log("msg", "Invalid buildID", "buildID", buildID)
+		return "", err
+	}
+
+	return sanitizedBuildID, nil
+}
+
+// createSymbolizationRequest creates a symbolization request for a mapping group
+func (s *Symbolizer) createSymbolizationRequest(binaryName, buildID string, mapping *googlev1.Mapping, locs []locToSymbolize) request {
+	req := request{
+		buildID:    buildID,
+		binaryName: binaryName,
+		locations:  make([]*location, len(locs)),
+	}
+
+	for i, loc := range locs {
+		req.locations[i] = &location{
+			address: loc.loc.Address,
+			mapping: &pprof.Mapping{
+				Start:   mapping.MemoryStart,
+				Limit:   mapping.MemoryLimit,
+				Offset:  mapping.FileOffset,
+				BuildID: buildID,
+			},
+		}
+	}
+
+	return req
+}
+
+// updateProfileWithSymbols updates the profile with symbolization results
+func (s *Symbolizer) updateProfileWithSymbols(profile *googlev1.Profile, mapping *googlev1.Mapping, locs []locToSymbolize, symLocs []*location) {
+	stringMap := make(map[string]int64, len(profile.StringTable))
+	for i, str := range profile.StringTable {
+		stringMap[str] = int64(i)
+	}
+
+	type funcKey struct {
+		nameIdx, filenameIdx int64
+	}
+	funcMap := make(map[funcKey]uint64)
+	maxFuncID := uint64(0)
```

Comment:
can we do `maxFuncId := len(profile.Function) + 1` here? I know I told you sometime ago it's not a valid approach in general, but it is a valid approach for pprof files returned from our pprof queries. so should be fine?

Reply:
hmm you're right. I've checked it and seems we do guarantee this so we can remove the loop to find `maxFuncID`, good one 👍